tests/xcm-simulator/src/experimental.rs

```diff
@@ -0,0 +1,235 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+use crate::mocks::{msg_queue::mock_msg_queue, parachain, *};
+
+use frame_support::{assert_ok, weights::Weight};
+use pallet_contracts::Determinism;
+use parity_scale_codec::{Decode, Encode};
+use sp_runtime::traits::Bounded;
+use xcm::{prelude::*, v3::Response};
+use xcm_simulator::TestExt;
+
+const GAS_LIMIT: Weight = Weight::from_parts(100_000_000_000, 3 * 1024 * 1024);
+
+#[test]
+fn basic_xcmp_outcome() {
+    MockNet::reset();
+
+    // basic remark call
+    let remark = parachain::RuntimeCall::System(
+        frame_system::Call::<parachain::Runtime>::remark_with_event {
+            remark: vec![1, 2, 3],
+        },
+    );
+
+    // priveleged root call
+    let root_call =
+        parachain::RuntimeCall::System(frame_system::Call::<parachain::Runtime>::set_storage {
+            items: vec![(vec![0], vec![1])],
+        });
+
+    let send_transcat = |call: parachain::RuntimeCall, dest: MultiLocation| {
+        let mut xcm = Xcm(vec![
+            WithdrawAsset((Here, 100_000_000_000_u128).into()),
+            BuyExecution {
+                fees: (Here, 100_000_000_000_u128).into(),
+                weight_limit: Unlimited,
+            },
+            Transact {
+                origin_kind: OriginKind::SovereignAccount,
+                require_weight_at_most: Weight::from_parts(1_000_000_000, 1024 * 1024),
+                call: call.encode().into(),
+            },
+            ExpectTransactStatus(MaybeErrorCode::Success),
+        ]);
+
+        // this will register the query and add `SetApendix` with `ReportError`.
+        ParachainPalletXcm::report_outcome(&mut xcm, dest, Bounded::max_value()).unwrap();
+        // We have to swap the appendix instruction with widthraw & buy execution
+        // to make barrier(AllowTopLevelPaidExecutionFrom) happy.
+        xcm.0.swap(0, 1);
+        xcm.0.swap(1, 2);
```

Comment:
I don't think it will help in this since `WithComputedOrigin` only accounts for `DescendOrigin` (also `UniversalOrigin`) and that's it.

The `AllowTopLevelPaidExecutionFrom` does not allow any instructions other than the ones putting asset in holding (and `ClearOrigin`)
https://github.com/paritytech/polkadot/blob/70a6e7e1e9351e989171d72a05ec0cd2a66a90db/xcm/xcm-builder/src/barriers.rs#L80-L93

Reply:
I checked the function you're using in more detail now.

I'd say it's not applicable for this scenario then.
If there's a _utility_ you use, and the very next thing you have to do is:
```
        xcm.0.swap(0, 1);
        xcm.0.swap(1, 2);
```
then it's probably not the right call to use it.

My guess is this is intended for system parachains/relay-chain communication, where the unpaid execution is allowed.

Reply:
Ok, let's use `new_query()` to register the query and add appendix instructions manually..